### PR TITLE
Allow Sagemaker training jobs resume an existing checkpoint

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#450 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by jalencato — COMMENTED** [review]
Overall comment, for training `--model-artifact-s3` is for saving model and for inference it is loading model, do we need to change the one in the inference to `--model-checkpoint-to-load` as well?

File: python/graphstorm/sagemaker/sagemaker_train.py [hunk 1]
```diff
@@ -213,6 +224,11 @@ def run_train(args, unknownargs):
         data_path, sagemaker_session)
     graph_config_path = download_graph(graph_data_s3, graph_name,
         host_rank, data_path, sagemaker_session)
+    if model_checkpoint_s3 is not None:
+        # Download Saved model checkpoint to resume
+        download_model(model_checkpoint_s3, restore_model_path, sagemaker_session)
+        print(f"{restore_model_path} {os.listdir(restore_model_path)}")
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/sagemaker_train.py:230]: Let's use logging instead of printing. `INFO` level would be fine here.

File: sagemaker/launch/launch_train.py [hunk 2]
```diff
@@ -134,6 +136,9 @@ def get_train_parser():
              "Do not store it with partitioned graph", required=True)
     training_args.add_argument("--model-artifact-s3", type=str, default=None,
         help="S3 bucket to save model artifacts")
+    training_args.add_argument("--model-checkpoint-to-load", type=str, default=None,
+        help="S3 bucket to a model checkpoint from a previous training task "
```
Comment (thvasilo) [inline sagemaker/launch/launch_train.py:140]:
```suggestion
        help="S3 path to a model checkpoint from a previous training task "
```

File: sagemaker/run/train_entry.py [hunk 3]
```diff
@@ -39,6 +39,9 @@ def parse_train_args():
              "Do not store it with partitioned graph")
     parser.add_argument("--model-artifact-s3", type=str,
         help="S3 location to store the model artifacts.")
+    parser.add_argument("--model-checkpoint-to-load", type=str, default=None,
+        help="S3 bucket to a model checkpoint from a previous training task "
```
Comment (thvasilo) [inline sagemaker/run/train_entry.py:43]:
```suggestion
        help="S3 path to a model checkpoint from a previous training task "
```

File: .github/workflow_scripts/lint_check.sh [hunk 4]
```diff
@@ -4,6 +4,7 @@ cd ../../
 set -ex
 
 python3 -m pip install --upgrade prospector pip
+yes | pip3 install astroid==v3.0.0
```
Comment (thvasilo) [inline .github/workflow_scripts/lint_check.sh:7]: Is this change related to this PR? If not let's do rebase to ensure only changes related to the PR show up.

File: docs/source/scale/sagemaker.rst [hunk 5]
```diff
@@ -160,6 +160,11 @@ Because we are using a three-partition OGB-MAG graph, we need to set the ``--ins
 
 The trained model artifact will be stored in the S3 location provided through the ``--model-artifact-s3`` argument. You can use the following command to check the model artifacts after the training completes.
 
+If you want to resume a saved model checkpoint to do model fine-tuning you can pass
+the s3 address of the model checkpoint through the ``--model-checkpoint-to-load``
```
Comment (thvasilo) [inline docs/source/scale/sagemaker.rst:164]:
```suggestion
the S3 address of the model checkpoint through the ``--model-checkpoint-to-load``
```

**Review by thvasilo — COMMENTED** [review]
Overall the changes look good. Let's rebase if possible to isolate the commit for this PR only.

Could you add some details about how this was tested?

**Comment by jalencato** [comment]
> Overall comment, for training `--model-artifact-s3` is for saving model and for inference it is loading model, do we need to change the one in the inference to `--model-checkpoint-to-load` as well?

After offline discussion, the original naming should be more reasonable.

File: .github/workflow_scripts/lint_check.sh (hunk 4 above)
Comment (classicsong) [reply .github/workflow_scripts/lint_check.sh:7]: With out this, there is a change the lint check will fail due to version miss-match between astroid and pylint.

**Review by jalencato — APPROVED** [review]

File: python/graphstorm/sagemaker/sagemaker_train.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/sagemaker/sagemaker_train.py:230]: Let's do it in another PR.

Comment (classicsong) [reply python/graphstorm/sagemaker/sagemaker_train.py:230]: We need to update all the scripts under sagemaker folder.

File: sagemaker/launch/launch_train.py (hunk 2 above)
Comment (classicsong) [reply sagemaker/launch/launch_train.py]: Update

File: sagemaker/run/train_entry.py (hunk 3 above)
Comment (classicsong) [reply sagemaker/run/train_entry.py]: Updated

File: docs/source/scale/sagemaker.rst (hunk 5 above)
Comment (classicsong) [reply docs/source/scale/sagemaker.rst]: Updated

**Review by thvasilo — APPROVED** [review]
LGTM
